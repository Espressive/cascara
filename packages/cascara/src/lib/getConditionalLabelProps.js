/**
 * Conditionally prepare label props in terms of the isLabeled flag:
 *
 * if true, the label is returned as id so it matches the htmlFor in rendered label tag,
 * otherwise, the label is returned as an aria-label so the node is still accessible
 *
 * @param {String} label The labelfor the node
 * @param {Boolean} isLabeled A flag that specifies if a label tag should be displayed
 * @returns {Object} conditionalLabelProps An object to be dot-spread'ed into the node
 */
export const getConditionalLabelProps = (label, isLabeled) =>
  isLabeled ? { id: label } : { 'aria-label': label };
