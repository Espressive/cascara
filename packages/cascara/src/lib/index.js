export { default as getStatusFromDataLength } from './getStatusFromDataLength';
