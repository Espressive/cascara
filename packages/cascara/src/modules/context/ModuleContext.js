import { createContext } from 'react';

// We should ONLY put values here that are needed for ALL Modules
const defaultValue = {
  isEditing: false,
  isEditingAny: false,
  value: 'blue',
};

/*
 * We will use this baseline context to create any of our context providers from.
 * Any unique functionality should be included in that provider and not in this Context.
 */
const ModuleContext = createContext(defaultValue);

/*
 * We are also exporting the default value in case we need to hydrate or merge
 * any providers with these default values.
 */
export { defaultValue };

export default ModuleContext;
