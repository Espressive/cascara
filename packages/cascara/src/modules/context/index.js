export { default as ModuleContext } from './ModuleContext';
export { default as ModuleProvider } from './ModuleProvider';
