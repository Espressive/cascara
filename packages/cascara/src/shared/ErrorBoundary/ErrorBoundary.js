import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import DefaultErrorTemplate from './templates/DefaultErrorTemplate';

/**
 * A ErrorBoundary will catch and display an error
 */

class ErrorBoundary extends PureComponent {
  static propTypes = {
    children: PropTypes.element,
    errorTemplate: PropTypes.func,
  };

  static defaultProps = {
    children: null,
    errorTemplate: DefaultErrorTemplate,
  };

  state = {
    error: null,
    errorInfo: null,
  };

  // Catch errors in any components below and re-render with error message
  // You can also log error messages to an error reporting service here
  componentDidCatch = (error, errorInfo) => {
    this.setState({
      error: error,
      errorInfo: errorInfo,
    });
  };

  render() {
    const { children, errorTemplate } = this.props;

    const { error, errorInfo } = this.state;

    if (errorInfo) {
      // Error path
      return React.createElement(errorTemplate, {
        error,
        errorInfo,
      });
    }
    // Normally, just render children
    return children;
  }
}

export default ErrorBoundary;
