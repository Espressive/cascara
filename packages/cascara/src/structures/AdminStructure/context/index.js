export { default as AdminContext } from './AdminContext';
export { default as AdminProvider } from './AdminProvider';
