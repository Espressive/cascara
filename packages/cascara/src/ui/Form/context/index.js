import FormProvider from './FormProvider';

export { FormProvider };
