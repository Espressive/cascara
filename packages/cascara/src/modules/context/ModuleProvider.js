import React from 'react';
import pt from 'prop-types';

import ModuleContext, { defaultValue } from './ModuleContext';
import useToggle from '../../hooks/useToggle';

const propTypes = {
  children: pt.oneOfType([pt.element, pt.arrayOf(pt.element)]),
  value: pt.shape({
    isEditing: pt.bool,
  }),
};

const ModuleProvider = ({ children, value, ...props }) => {
  const [isEditing, setIsEditing] = useToggle(value?.isEditing || false);

  const mergedValues = {
    ...defaultValue,
    isEditing,
    setIsEditing,
    ...value,
  };

  return (
    <ModuleContext.Provider value={mergedValues} {...props}>
      {children}
    </ModuleContext.Provider>
  );
};

ModuleProvider.propTypes = propTypes;

export default ModuleProvider;
