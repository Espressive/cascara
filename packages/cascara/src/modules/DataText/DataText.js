import React, { useContext } from 'react';
import { Input } from 'reakit/Input';
import pt from 'prop-types';
import { ModuleContext } from '../context';
import styles from '../DataModule.module.scss';

import { getConditionalLabelProps } from '../../lib/getConditionalLabelProps';
import ErrorBoundary from '../../private/ErrorBoundary';

const propTypes = {
  /** A module can have an Attribute, which will be used as form field name */
  attribute: pt.string,
  /** A Module can be defined to not present an editing state */
  isEditable: pt.bool,
  /** Presents the input without a label. NOT USER CONFIGURABLE */
  isLabeled: pt.bool,
  /** A Module needs to have a unique label relative to its context */
  label: pt.string,
  /** A Module can have a value */
  value: pt.string,
};

const DataText = ({
  attribute,
  isEditable = true,
  isLabeled = true,
  label,
  value,
  ...rest
}) => {
  const { isEditing, formMethods } = useContext(ModuleContext);

  const conditionalLabelProps = getConditionalLabelProps(label, isLabeled);

  const renderEditing = (
    <>
      {label && isLabeled && (
        <label htmlFor={attribute || label}>
          <span className={styles.LabelText}>{label || attribute}</span>
        </label>
      )}
      <Input
        {...conditionalLabelProps}
        {...rest}
        className={styles.Input}
        defaultValue={value}
        id={attribute || label}
        name={attribute || label}
        ref={formMethods?.register}
        type={'text'}
      />
    </>
  );

  const renderDisplay = (
    <span>
      {label && isLabeled && (
        <span className={styles.LabelText}>{label || attribute}</span>
      )}
      <span {...rest} aria-label={label} className={styles.Input}>
        {value}
      </span>
    </span>
  );

  // Do not render an editable input if the module is not editable
  return (
    <ErrorBoundary>
      <div className={styles.Text}>
        {isEditing && isEditable ? renderEditing : renderDisplay}
      </div>
    </ErrorBoundary>
  );
};

DataText.propTypes = propTypes;

export { propTypes };
export default DataText;
