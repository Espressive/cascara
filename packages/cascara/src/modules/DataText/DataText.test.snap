// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataText display snapshot 1`] = `
<div>
  <div
    class="Text"
  >
    <span>
      <span
        class="LabelText"
      >
        Display
      </span>
      <span
        class="Text"
      >
        my text value example
      </span>
    </span>
  </div>
</div>
`;

exports[`DataText editing snapshot 1`] = `
<div>
  <div
    class="Text"
  >
    <label
      for="Editing"
    >
      <span
        class="LabelText"
      >
        Editing
      </span>
      <input
        class="Input"
        id="Editing"
        name="Editing"
        type="text"
        value="my text value example"
      />
    </label>
  </div>
</div>
`;
